GAWB-1260 Fix isProtected on workspace list

breilly2 opened the pull request:
I was a little quick to approve PR #379. There were some problems that affected the isProtected flag on workspace list. These changes bring back some case classes that had been previously deleted. They are used to augment the response from Rawls by interpreting the dbGap realm to mean that the workspace is "protected" (it's not a pure passthrough). I've also updated the tests to make sure that this works correctly.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

breilly2 committed:
Fix isProtected on workspace list

breilly2 commented on src/main/scala/org/broadinstitute/dsde/firecloud/model/Workspace.scala:
```diff
@@ -34,6 +35,37 @@ object WorkspaceCreate {
 
 }
 
+case class UIWorkspaceResponse(
```
Some of the fields here now use case classes from RawlsModel.

breilly2 commented on src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala:
```diff
@@ -52,12 +72,14 @@ trait WorkspaceApiService extends HttpService with FireCloudRequestBuilding
             requireUserInfo() { _ =>
               entity(as[WorkspaceCreate]) { createRequest => requestContext =>
-                val extReq = Post(FireCloudConfig.Rawls.workspacesUrl, createRequest)
+                val extReq = Post(FireCloudConfig.Rawls.workspacesUrl, WorkspaceCreate.toWorkspaceRequest(createRequest))
```
This had to come back explicitly (also for clone) because there's no reason for the compiler to go looking for an implicit type conversion.

breilly2 commented on src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala:
```diff
@@ -52,12 +72,14 @@ trait WorkspaceApiService extends HttpService with FireCloudRequestBuilding
         pathEnd {
           requireUserInfo() { _ =>
-            passthrough(rawlsWorkspacesRoot, HttpMethods.GET)
+            mapHttpResponse(transformListWorkspaceRequest) {
```
This (and the singular GET) are not pure passthrough methods. They interpret the realm from Rawls into the isProtected property.

breilly2 commented on src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:
```diff
@@ -281,7 +281,7 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
           //it should already be returned as JSON but for some strange reason it's being returned as text/plain
           //here we take the plain text and force it to be json so we can get the test to work
-          assert(WorkspaceCreate.isProtected(entity.asString.parseJson.convertTo[WorkspaceResponse].workspace))
+          assert(entity.asString.parseJson.convertTo[UIWorkspaceResponse].workspace.get.isProtected)
```
This is testing GET, not post, so it needs to convert to UIWorkspaceResponse and not use WorkspaceCreate.

rushtong commented on src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:
```diff
@@ -412,10 +422,31 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
+      Post(clonePath, orchestrationRequest) ~> dummyUserIdHeaders("1234") ~> sealRoute(workspaceRoutes) ~> check {
+        rawlsServer.verify(request().withPath(clonePath).withMethod("POST").withBody(rawlsRequest.toJson.prettyPrint))
+        status should equal(Created)
+        responseAs[Workspace] should equal(response)
```
From slack discussion: `rawlsRequest` should start off with `AttributeName("library", "published") -> AttributeBoolean(true)`, and then that will be transformed inside the clone endpoint. At that point, the rawls stub should see `AttributeName("library", "published") -> AttributeBoolean(false)` instead.

Additionally, one could start off with an empty attribute map. The transform will populate the false value so that when Rawls sees the create request, it will use that value instead of whatever exists in the source workspace.

helgridly commented:
So from what I can tell, what you've done here is restored the UIWorkspace classes in order to translate "has NIH realm" to "is protected". I think my intention was to have the UI handle that instead, but I forgot to follow up on that.

My gut tells me that since Realms is going to be a more general concept, Orchestration should pass it through and let the UI deal with it. If the FireCloud (cancer-specific) UI wants to interpret that as "protected access if the realm is NIH" then it can do so. I think the answer depends on whether you see Orchestration as an NIH/cancer-y facade, or a passthrough/firewall service.

breilly2 commented:
That is correct. I guess it does depend on your point of view. In the absence of UI changes, I went with changing orchestration. We have some stories in our backlog to support multiple realms in the UI at which point the NIH realm and "protected" concept stops being a special case (or at least less of one).

I can add some comments in the code and relevant jira ticket to make sure we address this at some point. For now, I like having this here because of the test coverage it allows for with current techniques.

BTW, I think this is the first time I've heard anyone talk about Firecloud as being cancer-specific.

helgridly commented:
I'm okay with this reasoning. Let's tag the other TLs and see if there are any strong objections, though I doubt there will be. @davidangb @dvoet

MatthewBemis reviewed (approved):
I'm also cool with this in a restorative sense. Ideally I don't think that orch should care about specific realms at all- it would be the UI that decides to treat dbGap protected workspaces differently. But we can deal with that discussion when we start to build out the UI for realms. This gets us back to where we were before so :+1:

breilly2 committed:
Added comments and an additional test specifically for cloning a published workspace.

breilly2 commented:
I added some code comments, including notes about possibly moving the special case to firecloud-ui. Also added another test specifically for Greg's case of cloning a published workspace.

breilly2 committed:
Modified expectation for the test for cloning a published workspace.

breilly2 commented:
One more commit. I had to rebase to get @rushtong's latest changes because it affects a test that I added.
